CCE executable

moxcodes opened the pull request:
## Proposed changes

Puts the previous CCE utilities all together into a working executable that extracts waves from input H5 worldtube data.

### Types of changes:

- [ ] Bugfix
- [x] New feature
- [ ] Refactor

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Dependencies
- [x] #1993 
- [x] #2009

moxcodes committed:
Template LinearSolve on boundary prefix

moxcodes committed:
Adjust tags for executable

moxcodes committed:
Add mutator for DuJ

moxcodes committed:
Make scri tag lists more user-friendly

moxcodes committed:
Add action CalculateScriInputs

moxcodes committed:
Add pup for SpinWeighted

moxcodes committed:
Add full action list to Evolution component

moxcodes committed:
Change communication to work with executable

moxcodes commented:
rebased on develop.
Cheers!

kidder commented:
there are clang-tidy issues that need attention

kidder commented on src/Evolution/Systems/Cce/PreSwshDerivatives.hpp:
```diff
@@ -252,7 +252,7 @@ struct PreSwshDerivatives<Tags::Exp2Beta> {
 
   using return_tags = tmpl::list<Tags::Exp2Beta>;
-  using argument_tags = tmpl::append<tmpl::list<pre_swsh_derivative_tags>>;
+  using argument_tags = tmpl::append<pre_swsh_derivative_tags>;
```
drop the append

kidder commented on src/DataStructures/SpinWeighted.hpp:
```diff
@@ -202,11 +202,19 @@ struct SpinWeighted<T, Spin, true> {
 
+template <typename T, int Spin>
+void SpinWeighted<T, Spin, true>::pup(PUP::er& p) noexcept {
+  data_.pup(p);
```
why not `p | data_`

kidder commented on tests/Unit/DataStructures/Test_SpinWeighted.cpp:
```diff
@@ -211,6 +211,9 @@ SPECTRE_TEST_CASE("Unit.DataStructures.SpinWeighted",
   // check unchanged if no resize
   destructive_resize_check.destructive_resize(5);
   CHECK(destructive_resize_check == destructive_resize_copy);
+  const auto serialized_and_deserialized_copy =
```
can you test the case where the third template parameter is false?

kidder commented on src/Evolution/Systems/Cce/Components/CharacteristicEvolution.hpp:
```diff
@@ -73,12 +93,86 @@ struct CharacteristicEvolution {
+                     typename Metavariables::cce_boundary_component,
+                     CharacteristicEvolution<Metavariables>>,
+                 Actions::ReceiveWorldtubeData<Metavariables>,
+                 ::Actions::MutateApply<
```
we would prefer a new action or (incorporate into ReceiveWorldtubeData) that just combines all of these MutateApply's  into a single action.  Is there any reason to have them separate?

moxcodes replied:
No particular reason to split them. Would you be satisfied with a typelist? Given that the additional action would just be a sequence of `mutate_apply` calls it doesn't seem to add a whole lot aside from the puffy template code, documentation, and tests that would have to be included with a new action.

kidder commented on src/Evolution/Systems/Cce/Components/CharacteristicEvolution.hpp:
```diff
@@ -73,12 +93,86 @@ struct CharacteristicEvolution {
+          typename Metavariables::evolved_coordinates_variables_tag>,
+      ::Actions::UpdateU<::Tags::Variables<
+          tmpl::list<typename Metavariables::evolved_swsh_tag>>>,
+      ::Actions::MutateApply<GaugeUpdateAngularFromCartesian<
```
again combine MutateApply's into a single action

kidder reviewed (changes requested):
can you please split the commits 
- Add `get_value` overloads for new receive tags
- Add example cce yaml
- Add Cce Executable
- Add Cce code overview.
into a separate PR?

moxcodes commented:
fixups posted, among which are a typelist instead of a new action in the hopes that my reviewers will be appeased with the easier version of the changes :P . If it is insufficient, let me know and I'll go the long route.
Cheers!

moxcodes committed:
Fix clang-tidy warnings for VectorAlgebra

moxcodes commented:
Applying in progress as I apparently broke the build somehow

moxcodes commented:
A rebase seems to have improved matters

nilsvu commented on src/Evolution/Systems/Cce/OptionTags.hpp:
```diff
@@ -165,15 +163,34 @@ struct TargetStepSize : db::SimpleTag {
+  template <typename Metavariables>
+  static double create_from_options(double end_time,
+                                    const std::string& filename) {
+    if (end_time == std::numeric_limits<double>::infinity()) {
```
It would be helpful to have a comment here what happens when `EndTime` is infty. Also, could you add documentation to this tag? Not sure why the the other tags don't have documentation, but since this tag actually has a nontrivial implementation it would be useful to add it to the docs.

nilsvu commented on src/Evolution/Systems/Cce/Actions/CalculateScriInputs.hpp:
```diff
@@ -0,0 +1,95 @@
+  template <typename DbTags, typename... InboxTags, typename Metavariables,
+            typename ArrayIndex, typename ActionList,
+            typename ParallelComponent>
+  static auto apply(db::DataBox<DbTags>& box,
```
You can make the return value `std::tuple<db::DataBox<DbTags>&&>` and `return {std::move(box)}`. I think we have started to prefer this pattern.

nilsvu commented on src/Evolution/Systems/Cce/Actions/CalculateScriInputs.hpp:
```diff
@@ -0,0 +1,95 @@
+                     typename TagPack::derivative_of>&... arguments) noexcept {
+          Spectral::Swsh::angular_derivatives<
+              tmpl::list<typename TagPack::derivative_kind...>>(
+              l_max, 1, make_not_null(&get(*derivatives))...,
```
I find that it improves readability to factor literal arguments like this `1` into a variable so that it has a name and the reader understands what it means without looking up the function definition. You could do `static constexpr size_t num_radial_points = 1` and pass that to the function (just a suggestion).

nilsvu commented on tests/Unit/Evolution/Systems/Cce/Actions/Test_CalculateScriInputs.cpp:
```diff
@@ -0,0 +1,287 @@
+
+template <typename Metavariables>
+struct mock_characteristic_evolution {
+  using component_being_mocked = CharacteristicEvolution<Metavariables>;
```
If I understand correctly this does not actually mock any parallel component, so you can delete lines 71--73 and name this something like  `TestCharacteristicEvolution`. Mocking parallel components is only necessary when actions invoke other actions on other parallel components that are not present within the action testing setup, which I think is not the case here.

nilsvu commented on tests/Unit/Evolution/Systems/Cce/Actions/Test_CalculateScriInputs.cpp:
```diff
@@ -0,0 +1,287 @@
+  using compute_tags = db::AddComputeTags<>;
+
+  using initialize_action_list =
+      tmpl::list<ActionTesting::InitializeDataBox<simple_tags, compute_tags>>;
```
you can just inline this action list below in line 99 and also inline the `simple_tags` and `compute_tags`. I don't think you need the `initialization_tags` in line 90.

moxcodes replied:
I'll inline the `compute_tags`, but I think it's more readable to split the `simple_tags` into a separate tag list to avoid too many nested angle brackets.

nilsvu replied:
since `InitializeDataBox` defaults to an empty `ComputeTagsList` you don't even need that :)

nilsvu commented on tests/Unit/Evolution/Systems/Cce/Actions/Test_CalculateScriInputs.cpp:
```diff
@@ -0,0 +1,287 @@
+                             initialize_action_list>,
+      Parallel::PhaseActions<
+          typename Metavariables::Phase, Metavariables::Phase::Evolve,
+          tmpl::flatten<tmpl::list<
```
you can drop the `flatten`

nilsvu commented on tests/Unit/Evolution/Systems/Cce/Actions/Test_CalculateScriInputs.cpp:
```diff
@@ -0,0 +1,287 @@
+  runner.set_phase(metavariables::Phase::Initialization);
+  // run the initialization to get the values into the databox
+  // Parallel::printf("first action\n");
+  // ActionTesting::next_action<component>(make_not_null(&runner), 0);
```
remove

nilsvu commented on tests/Unit/Evolution/Systems/Cce/Actions/Test_CalculateScriInputs.cpp:
```diff
@@ -0,0 +1,287 @@
+       i < tmpl::size<extra_pre_swsh_derivative_scri_tags>::value +
+               tmpl::size<extra_precomputation_scri_tags>::value + 1;
+       ++i) {
+    ActionTesting::next_action<component>(make_not_null(&runner), 0);
```
It would help if you added a comment which actions are expected to be invoked by this loop. I actually don't quite understand why you invoke `next_action` this many times, since you only have 3 actions in the phase.

nilsvu commented on src/Evolution/Systems/Cce/Components/CharacteristicEvolution.hpp:
```diff
@@ -100,8 +194,7 @@ struct CharacteristicEvolution {
     auto& local_cache = *(global_cache.ckLocalBranch());
-    if (next_phase == Metavariables::Phase::RegisterWithObserver or
-        next_phase == Metavariables::Phase::Evolve) {
+    if (next_phase == Metavariables::Phase::Evolve) {
```
Why do you have this `if` statement, can't you just remove it? Phases that don't have an action list will just be skipped anyway

kidder replied:
that is not true, `execute_next_phase` is called in every phase except the initialization phase.
If all that is done is calling `start_phase`, that just broadcasts to the component which will then change to the next phase and execute its action list for that phase, which in the case would do nothing (except the broadcast) so removing the if is not wrong...

nilsvu commented on src/DataStructures/SpinWeighted.hpp:
```diff
@@ -212,7 +215,12 @@ struct SpinWeighted<T, Spin, true> {
+}
+
+template <typename T, int Spin>
+void SpinWeighted<T, Spin, false>::pup(PUP::er& p) noexcept {
```
can you combine the two `pup` functions into one by adding the bool as template parameter?

moxcodes replied:
I tried that first, but because the pup is individually declared in the template specializations, the definition must be similarly specialized. (doesn't build otherwise)

moxcodes commented:
more fixups posted. Removed all of the alterations that are unnecessary due to #2026 and replaced them with the far smaller commit `Adjust tags for executable`.
Cheers!

nilsvu commented on src/Evolution/Systems/Cce/OptionTags.hpp:
```diff
@@ -169,6 +169,23 @@ struct TargetStepSize : db::SimpleTag {
   }
 };
 
+struct EndTime : db::SimpleTag {
```
You have this tag now in both `InitializationTags` and `Tags` namespaces. You probably only need it once

kidder replied:
yes, only once; either keep it after initialization, or remove the `create_from_options` from the non-initialization tag

nilsvu commented on tests/Unit/Evolution/Systems/Cce/Actions/Test_CalculateScriInputs.cpp:
```diff
@@ -83,30 +79,29 @@ struct mock_characteristic_evolution {
+                         extra_precomputation_scri_tags,
+                         tmpl::bind<::Actions::MutateApply,
+                                    tmpl::bind<local_precompute, tmpl::_1>>>,
+                     Actions::CalculateScriInputs>>>;
```
Do you remember if the test passed before this fixup commit? If so, can you add checks that could have caught this issue?

moxcodes replied:
it did; this alteration doesn't change any of the execution of this test. This just causes the two `MutateApply` actions to be determined from the two typelists above `extra_pre_swsh_derivative_scri_tags` and `extra_precomputation_scri_tags` instead of manually determined. This is so that if we add more things to the scri computation input typelists, this test can be altered only by the corresponding additions to those above typelists, rather than finding everywhere those tags would get used implicitly.

nilsvu reviewed (commented):
I'm happy with squashing the fixups

kidder commented on src/Evolution/Systems/Cce/Components/CharacteristicEvolution.hpp:
```diff
@@ -89,15 +89,8 @@ struct CharacteristicEvolution {
-                     CharacteristicEvolution<Metavariables>>,
-                 Actions::ReceiveWorldtubeData<Metavariables>,
-                 ::Actions::MutateApply<
+  using mutate_apply_initialize_first_hypersurface =
```
no, we really want one action that does what all the chained MutateApply's do.

moxcodes replied:
Would you mind explaining why it is so strongly preferred? It has the cursory appearance of preferring code and test duplication.

kidder replied:
We prefer one action to do multiple things instead of chaining multiple actions that mutate one thing in the DataBox.    In general, if no communication is being done between actions, then those actions should just be combined into one action calling multiple functions.    One option could be to have a MutateApply that mutates multiple things, but one concern to watch out for is whether doing so causes you to put more things in the DataBox than you really need to put in.  In other words, if the chain you have has some initial inputs and computes some outputs, are there temporary quantities in the DataBox that only exist because you are mutating one thing at a time.

moxcodes replied:
I see, the suggested design is intended to avoid the buildup of temporaries in the databox? I suppose I could omit the `CauchyAngularCoords` from the databox in each of these chains in favor of allocating the tensor in both of the new actions and making the function calls more ugly by needing to specify all of the `apply` arguments explicitly.

So is it against the philosophy of this request to create an action `MutateApplyAll` taking a `tmpl::list` of invokables and just calls `mutate_apply` with each of the invokables in sequence on the databox?

nilsdeppe replied:
In general a `MutateApplyAll` is not ideal, but I'm fine doing that in the interest of getting this merged sooner rather than later :)

moxcodes replied:
Okay; do I understand correctly then that the real preference is toward larger actions fully specialized to their task; and if that task happens to be several `mutate_apply`s, then it should do that within its `apply`, but a long action list is undesirable?
I'm just genuinely trying to understand the current ideas for best-practices for action lists, because the request to split up the Initialization action in the previous PR seems at odds with a best-practice idea of keeping actions as bundled as can be practically achieved.

nilsdeppe replied:
Ah, okay, I see. Yes, for initialization actions we've found that a lot of them can be reused across different evolution systems or even shared with the elliptic solver. Additionally, since we only do initialization once, there's a lot less need for optimizations that are only possible when you know exactly what system you're solving, etc. Since the initialization actions also add things for the evolution, if temporaries are removed from the evolution, the associated initialization code also disappears.

For the evolution part I (we) think it's better to have fewer actions that do a lot of work, are optimized to reduce allocations while maintaining a small memory footprint, and are even specialized for systems. An example is for DG-subcell it'll be a lot easier to have a large RHS, slice, send, compute volume terms, etc. action than many small ones because the code is specialized to DG-subcell, and so won't generically be portable anyway, and we can write free functions if it happens to be portable.

I've generally also found that as code becomes more spread out and is only tied together in a super generic "Tie things" structure it becomes harder to follow control flow or to search for things in the code. This is a big reason why I think we should move to making some of these wrappers around wrappers go away.

Does that help clarify where we're coming from?

moxcodes replied:
Yeah, that helps. It's certainly true that most of the CCE hypersurface procedure is similarly non-generalizable.

nilsdeppe replied:
Yes, but I think it's better to deal with that sort of possible refactors after this is merged and some performance measurements can be made :)

kidder commented on src/Evolution/Systems/Cce/OptionTags.hpp:
```diff
@@ -261,6 +278,13 @@ struct StartTime : db::SimpleTag {
+/// either from option specification or from the file
+///
+/// \details If the option `OptionTags::EndTime` is set to
+/// `std::numeric_limits<double>::infinity()`, this will find the end time from
```
doxygen is complaining about `infinity`

kidder commented on src/Evolution/Systems/Cce/Actions/CalculateScriInputs.hpp:
```diff
@@ -83,10 +83,11 @@ struct CalculateScriInputs {
               tmpl::list<typename TagPack::derivative_kind...>>(
-              l_max, 1, make_not_null(&get(*derivatives))...,
-              get(arguments)...);
+              l_max, one_radial_point_for_boundary_derivative,
```
please revert this to 1 .   (this is why the documentation exists...)

kidder commented on src/Evolution/Systems/Cce/Components/CharacteristicEvolution.hpp:
```diff
@@ -205,11 +205,9 @@ struct CharacteristicEvolution {
-          local_cache)
-          .start_phase(next_phase);
-    }
+    Parallel::get_parallel_component<CharacteristicEvolution<Metavariables>>(
```
why was this change made?

moxcodes replied:
upon @nilsleiffischer 's request to remove the `if`

kidder reviewed (changes requested):
reviewed with @nilsdeppe

moxcodes committed:
Add action UpdateGauge

moxcodes committed:
Add action InitializeFirstHypersurface

moxcodes commented:
posted fixups and new commits for the addition of new actions. The commits since the previous review are:
`Add action UpdateGauge`
`Add action InitializeFirstHypersurface`
`fixup: OptionTags remove duplicate and dox fix`
`fixup: revert explicit var for 1 in CalculateScri`
Cheers!

kidder commented on src/Evolution/Systems/Cce/Components/CharacteristicEvolution.hpp:
```diff
@@ -165,24 +158,16 @@ struct CharacteristicEvolution {
       Actions::FilterSwshVolumeQuantity<Tags::BondiH>,
       compute_scri_quantities_and_observe, record_time_stepper_data_and_step,
-      mutate_apply_update_gauge, Actions::ExitIfEndTimeReached,
+      Actions::ExitIfEndTimeReached,
```
just curious why you moved when the action is called...

moxcodes replied:
When looking at this to create the new actions, I realized that the update gauge could be called only during `Evolve` if it was placed at the beginning -- the corresponding gauge update is not performed at the end of `Initialize` with this change, so fewer things needed to go into the `InitializeFirstHypersurface` action.

kidder reviewed (commented):
with @nilsdeppe 
you can squash; thanks!

moxcodes commented:
squashed.
Cheers!

nilsdeppe reviewed (approved):
![image](https://user-images.githubusercontent.com/13531030/76147809-fec13c80-606d-11ea-8c3d-7d42b5cd9c29.png)
